base/callback_helper.h

```diff
@@ -0,0 +1,42 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_BASE_CALLBACK_HELPER_H_
+#define BRAVE_BASE_CALLBACK_HELPER_H_
+
+#include <utility>
+
+#include "base/bind.h"
+#include "base/callback.h"
+#include "base/memory/scoped_refptr.h"
+#include "base/sequenced_task_runner.h"
+#include "base/task/post_task.h"
+
+namespace base {
+
+// Bind the OnceCallback and gurantee that it will be ran on certain sequence.
+// Normally we want callback to be ran on current
+// thread(base::SequencedTaskRunnerHandle::Get()) so no matter callback.Run() is
+// called on which sequence, it will alwasy be ran on the one we pass in.
+template <typename R, typename... Args>
+base::OnceCallback<R(Args...)> BindOnceCallbackToSequence(
+    scoped_refptr<base::SequencedTaskRunner> task_runner,
+    base::OnceCallback<R(Args...)> callback) {
+  return base::BindOnce(
+      [](scoped_refptr<base::SequencedTaskRunner> task_runner,
+         base::OnceCallback<R(Args...)> callback, Args... args) {
+        if (task_runner->RunsTasksInCurrentSequence())
```

Comment:
To be totally honest I find this helper more harmful, than useful, because it hides the nature of executed code. For the user it is absolutely unclear whether the callback is executed synchronously or posted, which is something really important. So I'm strongly opionated against using it as is.

And it's not really clear what problem it helps to solve in this given PR: if it's just needed to make a callback bound to the Factory's thread, there are handier alternatives like `media/base/bind_to_current_loop.h`

Reply:
This does exactly the same thing as BindToLoop, but I didn't know that existed an it's in `media` which is odd. Why are you strongly against this if you are suggesting using a nearly identical method?

Reply:
actually this is not the same as what I explained to @darkdh, I did want a parameter to pass in the task runner

Reply:
I suggested maybe also having `BindToCurrentLoop`, but wasn't sure if that was a good idea or not

Reply:
also it is always posted so that is not correct. The lambda runs async just like the original method would have. The only difference is that if it's on the wrong thread it will post to the correct thread

Reply:
> a nearly identical method?

I'm totally fine with anything similar, but it should be explicit about posting vs running a callback synchronously. So we shouldn't hide the check `if (task_runner->RunsTasksInCurrentSequence())` and do something depending on that

Reply:
fun fact: `media/base/bind_to_current_loop.h` is already copypasted to `components/sync/base/bind_to_current_task_runner.h` 🤦

Reply:
fixed in https://github.com/brave/brave-core/pull/7927/commits/1164a01075489add79d871f4a9a75be2808576eb using `base::BindPostTask`